### ISSUE TITLE
Plans to fusing matmul with xentropy?

### ISSUE THREAD

**Issue by yzhangcs** [mgmalek/efficient_cross_entropy#2]
@mgmalek Hi, great work!
Do you have any plans to further fuse the torch.matmul with xentropy inside one kernel to get rid of the loops?
I think it's a little bit more involved as you may need to utilize [super grouping](https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html) tricks to match the performance with GEMMs if necessary.

**Comment by YouJiacheng** [comment]
I think a CUDA graph capture is enough. <del>Accumulation with global memory (the `addmm`) seems unavoidable, fusion can only reduce the overhead of kernel launch.</del>

Hmmm, it seems that we COULD reduce the overhead of the global memory write for dlogits.
It seems that we can use fp32 accumulator for the gradient w.r.t. the weight to improve the numerical accuracy, without a fp32 global memory buffer. -- It seems impossible with current implementation, let me think...

**Comment by yzhangcs** [comment]
@YouJiacheng yea, if doing addmm outside, I think the overhead is inevitable. 
Also I think the benefits of current impls could be minor as we must allocate tensors for dw, for very large hidden size and relative small bsz x ctx, there's not much difference from the size of logits.
Maybe that's why flashattn does not provide such alg lol.
